Enable drivers for mimxrt595 evk

willisyi opened the pull request:
Enabled I2C, spi, dma, counter, wdog, flash storage for mimxrt595.

ChayGuo committed:
boards: Add I2C and sensor driver support on MIMXRT595-EVK

Enable I2C access to FXOS7000 sensor on the mixrt595_evk board
Tested using samples/sensor/fxos8700 for mimxrt595_evk_cm33.

Signed-off-by: Chay Guo <changyi.guo@nxp.com>

ChayGuo committed:
drivers: clock_control: mcux_lpc_syscon_clock: Add flexcomm16 suppport

Added FlexCOMM16(HS SPI1) clock for MIMXRT5xx.

Signed-off-by: Chay Guo <changyi.guo@nxp.com>

ChayGuo committed:
boards: arm: Add wdog support on mimxrt595_evk

Add watchdog support to the mimxrt595 platform.
The mimxrt595 platform is excluded from the watchdog
test case because the test case uses variables in the
noinit section that need to be retained through a reset
but the rt595 does not retain this memory through a
reset.

Signed-off-by: Chay Guo <changyi.guo@nxp.com>

ChayGuo committed:
boards: arm: Enable flash for storage on mimxrt595_evk

- The MX25UM51345G flash is connected to FLEXSPI PortA for
  mimxrt595_evk.
- Updated flexspi_mx25um51345g driver to support DTR OPI mode.
- Tested with tests/drivers/flash.

Signed-off-by: Chay Guo <changyi.guo@nxp.com>

gmarull commented on boards/arm/mimxrt595_evk/mimxrt595_evk_cm33-pinctrl.dtsi:
```diff
@@ -24,6 +24,23 @@
+			drive-strength = "high";
+			drive-open-drain;
+		};
+		group1 {
+			pinmux = <GPIO_PIO022_PIO0_22>;
+			input-enable;
+			slew-rate = "normal";
+			drive-strength = "normal";
+		};
```
why is this needed? sensor is configuring GPIO using the GPIO API.

ChayGuo replied:
It's indeed not needed. Removed it. Thanks.

mmahadevan108 commented on include/zephyr/dt-bindings/clock/mcux_lpc_syscon_clock.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020-22, NXP
+ * Copyright (c) 2020-2022, NXP
```
We can delete the `(c)` from the Copyright

ChayGuo replied:
Deleted the (c), thanks.

ChayGuo commented on samples/sensor/fxos8700/README.rst:
```diff
@@ -121,6 +121,18 @@ Sample can be built and executed for the MIMXRT685-EVK as follows:
 
+Building and Running for MIMXRT595-EVK
+======================================
+MIMXRT595-EVK is optional equipped with FXOS8700CQ accelerometer and magnetometer.
```
corrected, thanks.

ChayGuo committed:
boards: arm: Added DMA support on MIMXRT595-EVK

Added DMA driver support.
Tested with tests/drivers/dma/loop_transfer

Signed-off-by: Chay Guo <changyi.guo@nxp.com>

ChayGuo committed:
boards: mimxrt595: Add SPI support

Enable access to the HS_SPI pins(JP26) on the mimxrt595_evk board.
Using DMA mode, tested with spi_loopback testcase.

Signed-off-by: Chay Guo <changyi.guo@nxp.com>

ChayGuo committed:
boards: arm: mimxrt595: Add CTimer driver for mimxrt595_evk

Add counter support using CTimer for RT595.
Tested with samples/drivers/counter/alarm.

Signed-off-by: Chay Guo <changyi.guo@nxp.com>

ChayGuo commented on tests/drivers/spi/spi_loopback/boards/mimxrt595_evk_cm33.conf:
```diff
@@ -0,0 +1,9 @@
+#
+
+CONFIG_SPI_LOOPBACK_DRV_NAME="HS_LSPI_1"
+# CONFIG_SPI_MCUX_FLEXCOMM_DMA=y
```
Yes, tested with DMA. I've rebased the two commits for SPI and DMA enablement and split them to 1 for DMA support another for SPI support. 
Thanks.

ChayGuo commented on boards/arm/mimxrt595_evk/Kconfig.defconfig:
```diff
@@ -15,4 +15,11 @@ config FXOS8700_DRDY_INT1
 	default y
 	depends on FXOS8700_TRIGGER
 
+if DMA_MCUX_LPC
```
Accepted, thanks.

ChayGuo commented on soc/arm/nxp_imx/rt5xx/soc.h:
```diff
@@ -19,6 +19,9 @@
 #include <fsl_common.h>
 
+/* Add include for DTS generated information */
+#include <devicetree.h>
```
> e may not be needed. Please delete.

Removed. Thanks.

dleach02 commented:
@hakehuang can you test suite on this PR?

hakehuang commented:
> @hakehuang can you test suite on this PR?

@dleach02 , test is running will feedback once ready

gmarull commented on boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts:
```diff
@@ -104,6 +104,24 @@
+	compatible = "nxp,lpc-i2c";
+	status = "okay";
+	clock-frequency = <I2C_BITRATE_FAST>;
+	#address-cells = <1>;
+	#size-cells = <0>;
```
note: this should be moved to soc dts files

ChayGuo replied:
If I'm understanding correctly, the #address-cells, #size-cells properties are used for the child node "fxos8700@1e" which is defined below. So they are put here not in soc dts. 
Correct me if I'm wrong. 

Thanks. 
> note: this should be moved to soc dts files

gmarull commented on include/zephyr/dt-bindings/clock/mcux_lpc_syscon_clock.h:
```diff
@@ -23,6 +23,7 @@
 #define MCUX_FLEXCOMM13_CLK		13
 #define MCUX_HS_SPI_CLK			14
 #define MCUX_PMIC_I2C_CLK		15
+#define MCUX_HS_SPI1_CLK        16
```
nit: align using tabs as other entries

ChayGuo replied:
Fixed. Thanks. 

> nit: align using tabs as other entries

gmarull commented on boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.yaml:
```diff
@@ -21,3 +21,5 @@ supported:
   - gpio
   - i2c
+  - spi
+  - hs_spi1
```
where is this tag used?

ChayGuo replied:
Deleted it since it's not used currently.

gmarull reviewed (commented):
lgtm, a few minor comments

ChayGuo commented:
Updated the PR per review comments and resolved the conflicts in wdt_basic_api/testcase.yaml

nvlsianpu reviewed (commented):
flash driver changes looks fine to me.
Needs approval from platform maintainer,

mmahadevan108 commented on boards/arm/mimxrt595_evk/doc/index.rst:
```diff
@@ -77,6 +77,14 @@ features:
+| CTIMER    | on-chip    | counter                             |
++-----------+------------+-------------------------------------+
+| WDT       | on-chip    | watchdog                            |
++-----------+------------+-------------------------------------+
```
Add FLASH to this list

ChayGuo replied:
> Add FLASH to this list

Added. Thanks.

ChayGuo commented on drivers/flash/flash_mcux_flexspi_mx25um51345g.c:
```diff
@@ -36,6 +36,13 @@ static uint8_t nor_write_buf[SPI_NOR_PAGE_SIZE];
+#else
+#define NOR_FLASH_ENABLE_OCTAL_CMD 0x1
+#endif
+
```
> Can we avoid this platform specific define and use the same CMD for other platforms that use this flash device. Also is there a benefit to using `STR` over `DTR`

For RT685 EVK, the Octal flash is connected to Flexspi PortB, and no DQS pin is available, so if DTR used, the flexspi clock is restricted to <50MHz slower than STR(can be 100MHz), and RT685 ROM uses STR mode. 
For RT595 EVK the Octal flash uses DQS pin and can make higher performance using DTR mode, the ROM also uses DTR mode. 
This is platform/board-specific.

mmahadevan108 replied:
Then should this be a `Kconfig` that is enabled at the board level.

ChayGuo replied:
> Then should this be a `Kconfig` that is enabled at the board level.

Thanks, added a Kconfig choice in the flash kconfig file, and the value can be set at board level.
